pkg/products/marin3r/softLimitPrometheusRules.go

```diff
@@ -0,0 +1,51 @@
+package marin3r
+
+import (
+	"fmt"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/integr8ly/integreatly-operator/pkg/resources"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func (r *Reconciler) newSoftLimitAlertsReconciler() resources.AlertReconciler {
+	softDailyLimits := r.RateLimitConfig.SoftDailyLimits
+	if softDailyLimits == nil || len(softDailyLimits) == 0 {
+		return &resources.NoopAlertReconciler{}
+	}
+
+	return &resources.AlertReconcilerImpl{
+		ProductName:  "3scale",
+		Installation: r.installation,
+		Logger:       r.logger,
+		Alerts: []resources.AlertConfiguration{
+			{
+				AlertName: "rate-limit-soft-limits",
+				GroupName: "soft-limits.rules",
+				Namespace: r.Config.GetNamespace(),
+				Rules:     mapSoftLimitRules(softDailyLimits),
+			},
+		},
+	}
+}
+
+func mapSoftLimitRules(softDailyLimits []uint32) []monitoringv1.Rule {
+	result := make([]monitoringv1.Rule, len(softDailyLimits))
+
+	for i, softDailyLimit := range softDailyLimits {
+		result[i] = monitoringv1.Rule{
+			Alert: fmt.Sprintf("RHOAMApiUsageSoftLimitReachedTier%d", i+1),
+			Expr: intstr.FromString(
+				fmt.Sprintf("(ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits - ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits offset 1d) > %d", softDailyLimit),
```

Comment:
Just a question on this, as I'm not sure am I doing this right. I've been playing around with this `offset` modifier.
For me it looks to return no data if offset is in a range where there was no data present:

![image](https://user-images.githubusercontent.com/24636860/98667725-b8338880-2346-11eb-95bb-c495b7044aef.png)

But it will return data if there is data thoughout this range: 

![image](https://user-images.githubusercontent.com/24636860/98667910-0183d800-2347-11eb-9c92-ebf5fa2d5823.png)

So because of this, I'm thinking these alerts wont fire on the first day on install due to the above. The alert would fire on the second day once the expression can resolve correctly :thinking: 

![image](https://user-images.githubusercontent.com/24636860/98668367-b3bb9f80-2347-11eb-8ea9-2e38921b6310.png)

Wondering is this a concern or am I missing something ? :sweat_smile:

Reply:
No, you're right, I noticed this as well, that's why I suggest to modify it to 5 minutes in order to test the query. I'm not sure how much of a concern it'd be, as it means that if any of the limits is reached in the first day after the cluster is created it wouldn't trigger an alert. @pmccarthy wdyt?

Reply:
Would it work with `increase(ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits[24h]) > <value>`? We are using a similar query in the customer grafana dashboard.

Reply:
Yeah, i think using `increase` would work if this is a concern :thinking:

Reply:
If we have an alternative which would allow for the metric to be in place rather than not having it available on the first day, Í think we should go with that.

Reply:
@laurafitzgerald @KevFan I experimented with the `increase` function but for some reason the values I see in the graph don't make sense (they keep fluctuating even after the requests stop). I managed to come up with a query that defaults to 0 if there's no data. It's a bit complicated but seems to do the job. I'm not an expert in promql so there might be a better way. Wdyt? 

I broke down the query for readability, hope it makes sense:
```sql
# Current number of requests
scalar(ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits)
-
scalar(
  # Number of requests a day ago
  (ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits offset 1d) 
  or
  # Current number of requests - Current number of requests => 0 (Using vector(0) doesn't work as the vector types are different)
  (ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits - 
   ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits)
)
```

Reply:
Now this one looks a bit nicer
```sql
scalar(ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits) - 
scalar(
  (ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits offset 1d)
  or
  on() vector(0)
)
```